internal/consensus/pbts_test.go

```diff
@@ -182,51 +207,97 @@ func (p *pbtsTestHarness) nextHeight(proposer types.PrivValidator, deliverTime,
 	signAddVotes(p.ctx, p.observedState, tmproto.PrecommitType, p.chainID, bid, p.otherValidators...)
 	ensurePrecommit(p.t, p.ensureVoteCh, p.currentHeight, p.currentRound)
 
+	vk, err := p.observedValidator.GetPubKey(context.Background())
+	require.NoError(p.t, err)
+	res := collectHeightResults(p.ctx, p.t, p.eventCh, p.currentHeight, vk.Address())
+	ensureNewBlock(p.t, p.blockCh, p.currentHeight)
+
 	p.currentHeight++
 	incrementHeight(p.otherValidators...)
-	return <-p.resultCh
+	return res
 }
 
-// nolint: lll
-func registerResultCollector(ctx context.Context, t *testing.T, eb *eventbus.EventBus, address []byte) <-chan heightResult {
+func timestampedCollector(ctx context.Context, t *testing.T, eb *eventbus.EventBus) <-chan timestampedEvent {
 	t.Helper()
-	resultCh := make(chan heightResult, 2)
-	var res heightResult
+	eventCh := make(chan timestampedEvent, 100)
 	if err := eb.Observe(ctx, func(msg tmpubsub.Message) error {
-		ts := time.Now()
-		vote := msg.Data().(types.EventDataVote)
-		// we only fire for our own votes
-		if !bytes.Equal(address, vote.Vote.ValidatorAddress) {
-			return nil
-		}
-		if vote.Vote.Type != tmproto.PrevoteType {
-			return nil
+		eventCh <- timestampedEvent{
+			ts: time.Now(),
+			m:  msg,
 		}
-		res.prevoteIssuedAt = ts
-		res.prevote = vote.Vote
-		resultCh <- res
 		return nil
-	}, types.EventQueryVote); err != nil {
+	}, types.EventQueryVote, types.EventQueryCompleteProposal); err != nil {
 		t.Fatalf("Failed to observe query %v: %v", types.EventQueryVote, err)
 	}
-	return resultCh
+	return eventCh
+}
+
+// nolint: lll
+func collectHeightResults(ctx context.Context, t *testing.T, eventCh <-chan timestampedEvent, height int64, address []byte) heightResult {
+	t.Helper()
+	var res heightResult
+	for event := range eventCh {
+		switch v := event.m.Data().(type) {
+		case types.EventDataVote:
+			if v.Vote.Height > height {
+				t.Fatalf("received prevote from unexpected height, expected: %d, saw: %d", height, v.Vote.Height)
+			}
+			if !bytes.Equal(address, v.Vote.ValidatorAddress) {
+				continue
+			}
+			if v.Vote.Type != tmproto.PrevoteType {
+				continue
+			}
+			res.prevote = v.Vote
+			res.prevoteIssuedAt = event.ts
+
+		case types.EventDataCompleteProposal:
+			if v.Height > height {
+				t.Fatalf("received proposal from unexpected height, expected: %d, saw: %d", height, v.Height)
+			}
+			res.proposalIssuedAt = event.ts
+		}
+		if res.isComplete() {
+			break
+		}
+	}
+	return res
```

Comment:
Alternative suggestion:
```suggestion
		if res.isComplete() {
			return res
		}
	}
```

Rationale: The eye doesn't have to track past the boundary of the loop. 🙂

Reply:
Done, added a `panic` for the now unreachable code.

Reply:
It shouldn't be necessary, I think. The compiler should be able to see that this is statically unreachable.

Reply:
Oh, no, I'm wrong: The range can break.